MKR regression tests: flipper-bids-pass-rough, flopper-file-pass-rough, flopper-tick-pass-rough

ehildenb opened the pull request:
This updates the WordPack abstractions to remove the use of `smt-lemma`. Instead, equalities that look like `#lookup(_, _) ==Int WordPack(...)` are unfolded into their individual components, and there are lemmas provided for refolding them up when those storage locations are updated.

-   cat-ilks-pass is replaced with flipper-bids-pass-rough, which shows a bigger Solidity struct we need to handle hash non-overflow for.
-   flopper-file-pass-rough shows repacking a WordPackUInt48UInt48.
-   flopper-tick-pass-rough shows repacking a WordPackAddrUInt48UInt48.

daejunpark commented on tests/specs/mcd/verification.k:
```diff
@@ -23,23 +23,22 @@ module VERIFICATION-SYNTAX
-    rule maxUInt48  &Int (ADDR_UINT48_UINT48 /Int pow208) => UINT48_2 requires ADDR_UINT48_UINT48 ==Int #WordPackAddrUInt48UInt48(ADDR, UINT48_1, UINT48_2) andBool #rangeAddress(ADDR) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
+    rule UINT48_UINT48 ==K #WordPackUInt48UInt48(UINT48_1, UINT48_2) => UINT48_1 ==Int maxUInt48 &Int UINT48_UINT48 andBool UINT48_2 ==Int maxUInt48 &Int (UINT48_UINT48 /Int pow48)
+    rule (UINT48_2 *Int pow48) |Int (maskWordPackUInt48UInt48_1 &Int UINT48_UINT48) => #WordPackUInt48UInt48(maxUInt48 &Int UINT48_UINT48, UINT48_2                                 ) requires #rangeUInt(48, UINT48_2)
+    rule  UINT48_1             |Int (maskWordPackUInt48UInt48_2 &Int UINT48_UINT48) => #WordPackUInt48UInt48(UINT48_1                    , maxUInt48 &Int (UINT48_UINT48 /Int pow48)) requires #rangeUInt(48, UINT48_1)
```
No range condition for `UINT48_UINT48`?

daejunpark replied:
Also, it is hard to get the intuition of the above two lemmas.  Could you write a comment that explains them intuitively?

ehildenb replied:
I've added comments and re-organized the lemmas to hopefully be more intuitive.

daejunpark commented on tests/specs/mcd/verification.k:
```diff
@@ -23,23 +23,22 @@ module VERIFICATION-SYNTAX
 
-    rule maxUInt160 &Int  ADDR_UINT8              => ADDR  requires ADDR_UINT8 ==Int #WordPackAddrUInt8(ADDR, UINT8) andBool #rangeAddress(ADDR) andBool #rangeUInt(8, UINT8) [smt-lemma]
-    rule maxUInt8   &Int (ADDR_UINT8 /Int pow160) => UINT8 requires ADDR_UINT8 ==Int #WordPackAddrUInt8(ADDR, UINT8) andBool #rangeAddress(ADDR) andBool #rangeUInt(8, UINT8) [smt-lemma]
+    rule ADDR_UINT8 ==K #WordPackAddrUInt8(ADDR, UINT8) => ADDR ==Int maxUInt160 &Int ADDR_UINT8 andBool UINT8 ==Int maxUInt8 &Int (ADDR_UINT8 /Int pow160)
```
So, essentially you specified how to unify two terms when one of them is a WordPack term.  My concern is that this could be applied even if not necessary, leading to blowing up the size of constraints and serious performance issue.  Also, you may want to have a different unification strategy for some cases, e.g., `#WordPackAddrUInt8(X, Y) ==K #WordPackAddrUInt8(Z, W)`.

You can just go with these for now, but keep in mind the potential issue.

daejunpark replied:
BTW, no range condition for `ADDR_UINT8` needed?

ehildenb replied:
The re-organization of the lemmas and added comments should help make it clearer. Basically `_ ==K #WordPack(_, _)` itself is a large predicate which unfolds into several smaller predicates, which makes it easier to reason about the individual values packed into the storage slot.

denis-bogdanas commented on tests/specs/mcd/verification.k:
```diff
@@ -252,4 +251,9 @@ module VERIFICATION
 
+    // ### dsvalue-peek-pass-rough
+
+    rule (B ==K false) ==K false => B
```
On Haskell this is not needed: https://github.com/kframework/evm-semantics/blob/master/tests/specs/functional/lemmas-no-smt-spec.k#L37
Because of this: https://github.com/kframework/evm-semantics/blob/fc44b9ddd1ba29f433778303e922ea73ab748df2/tests/specs/lemmas.k#L268-L269

Lemmas don't have side conditions, so they don't carry performance degradation risks. Should be made common for Java & Haskell.

denis-bogdanas commented on tests/specs/mcd/verification.k:
```diff
@@ -252,4 +251,9 @@ module VERIFICATION
+    // ### dsvalue-peek-pass-rough
+
+    rule (B ==K false) ==K false => B
+    rule bool2Word((UINT8 ==K 0) ==K false) => UINT8 requires #rangeBool(UINT8)
```
With lemmas from above comment, this can be replaced with lower-level lemmas:
```
rule notBool (X ==Int 0) => X ==Int 1 requires #rangeBool(X)
rule bool2Word(X ==Int 1)  => X         requires #rangeBool(X)

//Rcommended for symmetry, to avoid dealing with them next time
rule notBool (X ==Int 1) => X ==Int 0   requires #rangeBool(X)
```
And above rule should be made test instead.

ehildenb commented:
I've removed `bool2Word` changes into a separate PR (https://github.com/kframework/evm-semantics/pull/918) so that they can be reviewed independently. This one has the refactored `WordPack` lemmas, which basically amounts to rules which (i) change predicates about `#WordPack` into predicates about the individual components, and (ii) fold updates to individual components back up into `#WordPack` over the entire word.

kmbarry1 commented on tests/specs/mcd/flipper-bids-pass-rough-spec.k:
```diff
@@ -0,0 +1,233 @@
+     andBool #lookup(ACCT_ID_ORIG_STORAGE, #Flipper.bids[ABI_n].usr) ==Int Junk_3
+     andBool #lookup(ACCT_ID_ORIG_STORAGE, #Flipper.bids[ABI_n].gal) ==Int Junk_4
+     andBool #lookup(ACCT_ID_ORIG_STORAGE, #Flipper.bids[ABI_n].tab) ==Int Junk_5
+     andBool #Flipper.bids[ABI_n].bid =/=Int #Flipper.bids[ABI_n].lot
+     andBool #Flipper.bids[ABI_n].bid =/=Int #Flipper.bids[ABI_n].guy_tic_end
+     andBool #Flipper.bids[ABI_n].bid =/=Int #Flipper.bids[ABI_n].usr
+     andBool #Flipper.bids[ABI_n].bid =/=Int #Flipper.bids[ABI_n].gal
+     andBool #Flipper.bids[ABI_n].bid =/=Int #Flipper.bids[ABI_n].tab
+     andBool #Flipper.bids[ABI_n].lot =/=Int #Flipper.bids[ABI_n].guy_tic_end
+     andBool #Flipper.bids[ABI_n].lot =/=Int #Flipper.bids[ABI_n].usr
+     andBool #Flipper.bids[ABI_n].lot =/=Int #Flipper.bids[ABI_n].gal
+     andBool #Flipper.bids[ABI_n].lot =/=Int #Flipper.bids[ABI_n].tab
+     andBool #Flipper.bids[ABI_n].guy_tic_end =/=Int #Flipper.bids[ABI_n].usr
+     andBool #Flipper.bids[ABI_n].guy_tic_end =/=Int #Flipper.bids[ABI_n].gal
+     andBool #Flipper.bids[ABI_n].guy_tic_end =/=Int #Flipper.bids[ABI_n].tab
+     andBool #Flipper.bids[ABI_n].usr =/=Int #Flipper.bids[ABI_n].gal
+     andBool #Flipper.bids[ABI_n].usr =/=Int #Flipper.bids[ABI_n].tab
+     andBool #Flipper.bids[ABI_n].gal =/=Int #Flipper.bids[ABI_n].tab
```
A bit surprised this storage address pairwise disequality check is necessary...shouldn't these all be of the form `hash + offset` where `hash` stays the same but `offset` is different for every slot? Not saying there's anything wrong with it, just surprised `K` doesn't recognize that they aren't the same.

ehildenb replied:
I started by adding lemmas about the disequalities of hashes on different inputs and at different offsets, but a discussion in #902 with @daejunpark made me realize it may be easy to add lemmas which _seem_ OK, but are not.

So I switched to generating exactly the disequalities needed in the proofs in #904 (and made KLab generate these disequalities directly on my branch `pr` of my fork). Because this (i) didn't seem to impact performance at all, (ii) was easy to make KLab do correctly, and (iii) expresses _exactly_ the disequalities needed for a given proof rather than the over-approximation given by the hash2 lemmas, it seemed like the better way to go.

ehildenb replied:
Specifically if you want to see the type of `hash2` disequalities that were removed, check out: https://github.com/kframework/evm-semantics/pull/902/commits/7d81a7bcaca9eb6c107566ebab6b1c99feada14b

daejunpark commented on tests/specs/mcd/verification.k:
```diff
@@ -19,27 +19,54 @@ module VERIFICATION-SYNTAX
+    rule    ADDR_UINT8 ==K #WordPackAddrUInt8(ADDR, UINT8)
+         => ADDR  ==Int maxUInt160 &Int  ADDR_UINT8
+    andBool UINT8 ==Int maxUInt8   &Int (ADDR_UINT8 /Int pow160)
+    andBool #rangeUInt(168, ADDR_UINT8)
```
Mentioning that the arguments are packed in the reverse order would help others to understand these rules.

ehildenb replied:
I now mention it at the beginning of the WordPack explanation comment.

daejunpark commented on tests/specs/mcd/verification.k:
```diff
@@ -19,27 +19,54 @@ module VERIFICATION-SYNTAX
+    syntax Int ::= "maskWordPackUInt48UInt48_1" | "maskWordPackUInt48UInt48_2"
+ // --------------------------------------------------------------------------
+    rule maskWordPackUInt48UInt48_1 => 115792089237316195423570985008687907853269984665561335876943319951794562400255 [macro] // pow256 -Int pow96 +Int pow48 -Int 1
+    rule maskWordPackUInt48UInt48_2 => 115792089237316195423570985008687907853269984665640564039457583726438152929280 [macro] // pow256 -Int pow48
```
Having these hex numbers in the comment would be helpful to understand.
```
// maskWordPackUInt48UInt48_1: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF000000000000FFFFFFFFFFFF
// maskWordPackUInt48UInt48_2: 0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF000000000000
```

ehildenb replied:
Added.

daejunpark commented on tests/specs/mcd/verification.k:
```diff
@@ -19,27 +19,54 @@ module VERIFICATION-SYNTAX
-    rule maxUInt48 &Int  UINT48_UINT48             => UINT48_1 requires UINT48_UINT48 ==Int #WordPackUInt48UInt48(UINT48_1, UINT48_2) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
-    rule maxUInt48 &Int (UINT48_UINT48 /Int pow48) => UINT48_2 requires UINT48_UINT48 ==Int #WordPackUInt48UInt48(UINT48_1, UINT48_2) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2) [smt-lemma]
+    rule  UINT48_1             |Int (maskWordPackUInt48UInt48_2 &Int UINT48_UINT48) => #WordPackUInt48UInt48( UINT48_1 , (maskWordPackUInt48UInt48_2 &Int UINT48_UINT48) /Int pow48 ) requires #rangeUInt(256, UINT48_UINT48) andBool #rangeUInt(48, UINT48_1)
+    rule (UINT48_2 *Int pow48) |Int (maxUInt48                  &Int UINT48_UINT48) => #WordPackUInt48UInt48( maxUInt48 &Int UINT48_UINT48 , UINT48_2                               ) requires #rangeUInt(256, UINT48_UINT48) andBool #rangeUInt(48, UINT48_2)
```
Isn't it `#rangeUInt(96, UINT48_UINT48)`?

The second rule might be fine even if the higher bits are non-zero, but the first rule would _not_.  The higher bits are not cleared by ANDing with `maskWordPackUInt48UInt48_2`.

daejunpark replied:
BTW, these rules are quite confusing mainly because of the reverse order of the arguments being packed.  Would be the spec hard to read if we reverse the order of the arguments in the WordPack* terms?

ehildenb replied:
Good catch, I've changed it to `#rangeUInt(96, UINT48_UINT48)`.

I would prefer not to change the argument order for the WordPack, because that's the order the KLab currently is generating them in. It's not huge change to make for it, but would rather focus on getting more proofs passing.

If you really want me to change the argument order (instead of just leaving the comment about it), let me know.

daejunpark commented on tests/specs/mcd/verification.k:
```diff
@@ -19,27 +19,54 @@ module VERIFICATION-SYNTAX
-    rule maxUInt160 &Int  ADDR_UINT8              => ADDR  requires ADDR_UINT8 ==Int #WordPackAddrUInt8(ADDR, UINT8) andBool #rangeAddress(ADDR) andBool #rangeUInt(8, UINT8) [smt-lemma]
-    rule maxUInt8   &Int (ADDR_UINT8 /Int pow160) => UINT8 requires ADDR_UINT8 ==Int #WordPackAddrUInt8(ADDR, UINT8) andBool #rangeAddress(ADDR) andBool #rangeUInt(8, UINT8) [smt-lemma]
+    rule (UINT48_2 *Int pow208) |Int (maxUInt208 &Int ADDR_UINT48_UINT48) => #WordPackAddrUInt48UInt48(maxUInt160 &Int ADDR_UINT48_UINT48, maxUInt48 &Int (ADDR_UINT48_UINT48 /Int pow160), UINT48_2) requires #rangeUInt(256, ADDR_UINT48_UINT48) andBool #rangeUInt(48, UINT48_2) andBool maxUInt48 &Int (ADDR_UINT48_UINT48 /Int pow160) =/=Int 0
+    rule (UINT48_2 *Int pow208) |Int (maxUInt208 &Int ADDR_UINT48_UINT48) => #WordPackAddrUInt48UInt48(maxUInt160 &Int ADDR_UINT48_UINT48, 0                                              , UINT48_2) requires #rangeUInt(256, ADDR_UINT48_UINT48) andBool #rangeUInt(48, UINT48_2) andBool maxUInt48 &Int (ADDR_UINT48_UINT48 /Int pow160)  ==Int 0
```
Could you please explain why this separation is needed?

ehildenb replied:
I added a comment/TODO explaining why this is needed on the Java backend.

daejunpark commented on tests/specs/mcd/verification.k:
```diff
@@ -19,27 +19,54 @@ module VERIFICATION-SYNTAX
  // -----------------------------------------------------------------------------------------------------------------
-    // rule #WordPackUInt48UInt48     (            UINT48_1 , UINT48_2 ) => UINT48_2 *Int pow48 +Int UINT48_1                        requires #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2)
-    // rule #WordPackAddrUInt48UInt48 (     ADDR , UINT48_1 , UINT48_2 ) => UINT48_2 *Int pow208 +Int UINT48_1 *Int pow160 +Int ADDR requires #rangeAddress(ADDR) andBool #rangeUInt(48, UINT48_1) andBool #rangeUInt(48, UINT48_2)
-    // rule #WordPackAddrUInt8        (     ADDR , UINT8               ) => UINT8 *Int pow160 +Int ADDR                              requires #rangeAddress(ADDR) andBool #rangeUInt(8, UINT_8)
```
I think it would be good to keep these commented rules for the documentation purpose, as they give the "definition" of the WordPack* abstraction, as well as their definedness conditions.

ehildenb replied:
Added them back.

daejunpark reviewed (commented):
Can we have some tests for the WordPack lemmas, as you did before for other tricky lemmas?

ehildenb commented:
@daejunpark I've addressed all your comments, except about adding tests. The tests that are added with this PR will not pass without these additional lemmas, and there are maybe 10 other MKR proofs which also won't pass with these lemmas. If they are bumped up to `tests/specs/work-pack.k` (so they can be used elsewhere), then I'm happy to add unit-tests for them, but for now I think the MKR integration tests should suffice.
